dipy/align/imaffine.py

```diff
@@ -0,0 +1,1010 @@
+            self.transformed = self.transformed.astype(np.float64)
+        else:
+            self.update_pdfs = self.update_pdfs_sparse
+            self.update_gradient = self.update_gradient_sparse
+            static_32 = np.array(static).astype(np.float32)
+            moving_32 = np.array(moving).astype(np.float32)
+            self.transformed = None
+            k = int(np.ceil(1.0 / self.sampling_proportion))
+            shape = np.array(static.shape, dtype=np.int32)
+            self.samples = sample_domain_regular(k, shape, static_grid2world)
+            self.samples = np.array(self.samples)
+            self.ns = self.samples.shape[0]
+            # Add a column of ones (homogeneous coordinates)
+            self.samples = np.hstack((self.samples, np.ones(self.ns)[:, None]))
+            # Sample the static image
+            static_p = self.static_world2grid.dot(self.samples.T).T
+            static_p = static_p[..., :self.dim]
+            self.static_vals, inside = self.interp_method(static_32, static_p)
+            self.static_vals = np.array(self.static_vals, dtype=np.float64)
+            # Sample the moving image
+            sp_to_moving = self.moving_world2grid.dot(P)
+            moving_p = sp_to_moving.dot(self.samples.T).T
+            moving_p = moving_p[..., :self.dim]
+            self.moving_vals, inside = self.interp_method(moving_32,
+                                                          moving_p)
+
+        MattesBase.setup(self, self.static, self.moving)
+
+    def _update(self, params, update_gradient=True):
+        r""" Updates marginal and joint distributions and the joint gradient
+
+        The distributions are updated according to the static and transformed
+        images. The transformed image is precisely the moving image after
+        transforming it by the transform defined by the `params` parameters.
+
+        The gradient of the joint PDF is computed only if update_gradient
+        is True.
+
+        Parameters
+        ----------
+        params : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+        update_gradient : Boolean, optional
+            if True, the gradient of the joint PDF will also be computed,
+            otherwise, only the marginal and joint PDFs will be computed.
+            The default is True.
+
+        Returns
+        -------
+        is_invertible : Boolean
+            whether the transform defined by `params` is invertible or not
+        """
+        # Get the matrix associated with the params parameter vector
+        M = self.transform.param_to_matrix(params)
+        if self.starting_affine is not None:
+            M = M.dot(self.starting_affine)
+        try:
+            i = npl.inv(M)
+        except:
+            return False
```

Comment:
>This seems like it is an error condition? Would it be sensible to raise an error instead?

Now raising a ValueError. We need to return some really bad value (np.inf) to the Optimizer instead of rising an exception, though. Otherwise, the Optimizer would stop after attempting to evaluate the metric at a non-invertible transform.